### PR TITLE
Added support for configurable socket address in filesystem for hot restart

### LINKED ISSUE #2908: Hot restart should support configurable socket address
Hot restart currently uses a hard-coded address in the abstract namespace for unix domain sockets. This has the advantage that it allows hot-restart between different containers. However, it has the disadvantage that there are no permissions on abstract namespace unix domain sockets.

Envoy should support configuring a filesystem path for it's hot-restart UDS so that normal filesystem permissions can be used. This would most-likely be specified as a command-line argument.

*Relevant Links*:
http://man7.org/linux/man-pages/man7/unix.7.html

### PR DESCRIPTION
**Pull request by mk46**
Signed-off-by: Manish Kumar <manish.kumar1@india.nec.com>

 
Commit Message: Added support for configurable socket address in the filesystem
Additional Description: Socket address can be configured by --socket-path which is the directory where the sock file created. if not specified then the socket will be created in abstract namespace. Its permission can be set by --socket-mode which takes mode_t type as input.  
 Fixes #2908

### CODE REVIEW

File: source/server/hot_restarting_base.cc
Comment (mk46) [reply source/server/hot_restarting_base.cc]: Done

File: source/server/hot_restarting_base.cc
Comment (mk46) [reply source/server/hot_restarting_base.cc:40]: Yes, This is valid.

File: source/server/options_impl.cc
Comment (mk46) [reply source/server/options_impl.cc]: Deleted octalToDecimal

File: source/server/options_impl.cc
Comment (ggreenway) [inline source/server/options_impl.cc:166]: I think a better default is `600`

File: source/server/options_impl.cc
Comment (ggreenway) [inline source/server/options_impl.cc:273]: This cast isn't safe; the callee could overwrite adjacent memory. Use a temporary uint64_t and then assign to `socket_mode_`.

File: docs/root/operations/cli.rst
Comment (ggreenway) [inline docs/root/operations/cli.rst:199]: Please also document that all envoy processes wanting to participate in hot-restart together must use the same value for this option.

File: source/server/options_impl.cc
Comment (mk46) [reply source/server/options_impl.cc]: Changed to 600

File: source/server/options_impl.cc
Comment (mk46) [reply source/server/options_impl.cc]: Done

File: docs/root/operations/cli.rst
Comment (mk46) [reply docs/root/operations/cli.rst]: Added in docs.

File: source/server/options_impl.cc
Comment (ggreenway) [inline source/server/options_impl.cc:275]: Does something fail if you leave it as `600` for an abstract namespace socket? If not, I think you can just omit this. Another idea is to construct the `PipeInstance` here, which includes a mode, so that you only have to pass one argument through the several function calls, instead of both address and mode.

Comment (mk46) [reply source/server/options_impl.cc:275]: Yes, It's failing. `PipeInstance` need 0 as mode if it's in abstract namespace otherwise throw an exception like https://github.com/envoyproxy/envoy/blob/02e97e1dcd948c14b62de95b348badb5ee6d1575/source/common/network/address_impl.cc#L280

Comment (ggreenway) [reply source/server/options_impl.cc:275]: Got it. It's fine to leave as-is then.

File: source/server/options_impl.cc
Comment (mk46) [reply source/server/options_impl.cc]: I think  socket_mode must be of mode_t type because for string type if the input is other than an integer will be accepted which will not desirable.WDYT?

Comment (ggreenway) [reply source/server/options_impl.cc]: The return value of `atoull` tells you if it was successfully converted to a number. If it fails, throw a `MalformedArgvException`

**Comment by mk46** [comment]
@ggreenway PTAL.

File: test/integration/hotrestart_test.sh
Comment (ggreenway) [inline test/integration/hotrestart_test.sh:283]: `${TEST_TMPDIR}/envoy_domain_socket`

File: test/server/options_impl_test.cc
Comment (ggreenway) [inline test/server/options_impl_test.cc:127]: I think you can use `0644`, which C will interpret as octal (because of leading 0), instead of `420`

File: test/server/options_impl_test.cc
Comment (ggreenway) [inline test/server/options_impl_test.cc:188]: `0644`

File: test/server/options_impl_test.cc
Comment (ggreenway) [inline test/server/options_impl_test.cc:218]: `0644`

File: test/integration/hotrestart_test.sh
Comment (mk46) [reply test/integration/hotrestart_test.sh:305]: Exceeds maximum UDS path size.

File: test/server/options_impl_test.cc
Comment (mk46) [reply test/server/options_impl_test.cc]: Done

File: test/server/options_impl_test.cc
Comment (mk46) [reply test/server/options_impl_test.cc]: Done

File: test/server/options_impl_test.cc
Comment (mk46) [reply test/server/options_impl_test.cc]: Done

File: source/server/options_impl.cc
Comment (mk46) [reply source/server/options_impl.cc]: Removed check for the non-number character.

File: docs/root/operations/cli.rst
Comment (ggreenway) [inline docs/root/operations/cli.rst:208]: nit: `This flag may not be used...`

File: docs/root/operations/cli.rst
Comment (ggreenway) [inline docs/root/operations/cli.rst:207]: Instead of giving a numeric range, I think it's clearer to say `This must be a valid octal file permission, such as 644. The default value is 600.`

File: docs/root/operations/cli.rst
Comment (ggreenway) [inline docs/root/operations/cli.rst:198]: /s/Prefix/Suffix/

File: test/integration/hotrestart_test.sh
Comment (ggreenway) [inline test/integration/hotrestart_test.sh:287]: I think it's probably good enough to test the new mode with only 1 of these configs; no need to re-run for all variants.

**Review by ggreenway — CHANGES_REQUESTED** [review]
This looks really good. Just a few more minor things.

**Comment by ggreenway** [comment]
Note: coverage test failure was due to disk-full in CI.

/wait

File: test/integration/hotrestart_test.sh
Comment (ggreenway) [reply test/integration/hotrestart_test.sh:305]: I'm worried that this may cause conflicts between concurrently running variations of the test. @envoyproxy/maintainers any guidance on this?

File: docs/root/operations/cli.rst
Comment (mk46) [reply docs/root/operations/cli.rst]: Done.

File: docs/root/operations/cli.rst
Comment (mk46) [reply docs/root/operations/cli.rst]: Done.

File: docs/root/operations/cli.rst
Comment (mk46) [reply docs/root/operations/cli.rst]: Done.

File: test/integration/hotrestart_test.sh
Comment (mk46) [reply test/integration/hotrestart_test.sh]: Done.

File: test/integration/hotrestart_test.sh
Comment (jmarantz) [inline test/integration/hotrestart_test.sh:287]: real symbol tables are the default, so I'd run just one with fake symbol tables (switch the "0" and "1" in the second arg in the two run_testsuite calls below).

File: test/integration/hotrestart_test.sh
Comment (jmarantz) [reply test/integration/hotrestart_test.sh:302]: agreed. Try this with --runs_per_test=100.

You could also make this unique I think using BASE_ID_PATH; not sure if that would exceed the limit. What is the limit?

Comment (mk46) [reply test/integration/hotrestart_test.sh:302]: Its limit is 108 bytes.

File: test/integration/hotrestart_test.sh
Comment (mk46) [reply test/integration/hotrestart_test.sh:287]: Done.

File: test/integration/hotrestart_test.sh
Comment (mk46) [reply test/integration/hotrestart_test.sh]: Done.

**Review by ggreenway — APPROVED** [review]
LGTM. Thanks for this!

**Comment by mk46** [comment]
@envoyproxy/api-shepherds Would you mind to take a look at this PR? Thanks.

**Comment by ggreenway** [comment]
@envoyproxy/api-shepherds

**Review by ggreenway — CHANGES_REQUESTED** [review]
Sorry, one last thing before we merge: need to add a release note in `docs/root/version_history/current.rst`.

**Comment by mk46** [comment]
@ggreenway  Added release note.  Would you mind to take a look?  Thanks.

File: docs/root/version_history/current.rst
Comment (ggreenway) [inline docs/root/version_history/current.rst:83]: Only need a single line/entry for the feature; maybe add a `:ref:` to cli.rst

**Review by ggreenway — CHANGES_REQUESTED** [review]

File: docs/root/version_history/current.rst
Comment (mk46) [reply docs/root/version_history/current.rst:83]: ` * hot restart: added :option:`--socket-path` and :option:`--socket-mode` to configure UDS path in the filesystem and set permission to it.`

Make sense?

Comment (ggreenway) [reply docs/root/version_history/current.rst:83]: Yes, that sounds good.

**Comment by ggreenway** [comment]
You have a merge conflict.

Please don't force-push to the PR branch. It makes it very difficult to tell what changed. Merges are fine though.

**Comment by mk46** [comment]
@ggreenway Thanks for your suggestion. I will TC from next time.

**Review by ggreenway — APPROVED** [review]

**Pull request merged**
